pkg/repl/client.go

```diff
@@ -395,18 +391,23 @@ func (c *Client) Close() {
 	}
 }
 
-func (c *Client) updatePosInMemory(pos uint32) {
-	c.Lock()
-	defer c.Unlock()
-	c.binlogPosInMemory = mysql.Position{
-		Name: c.lastLogFileName,
-		Pos:  pos,
-	}
-}
-
 // FlushUnderLock is a final flush under an exclusive lock using the connection
-// that holds a write lock.
+// that holds a write lock. Because flushing generates binary log events,
+// we actually want to call flush *twice*:
+//   - The first time flushes the pending changes to the new table.
+//   - We then ensure that we have all the binary log changes read from the server.
+//   - The second time reads through the changes generated by the first flush
+//     and updates the in memory applied position to match the server's position.
+//     This is required to satisfy the binlog position is updated for the c.AllChangesFlushed() check.
 func (c *Client) FlushUnderLock(ctx context.Context, lock *dbconn.TableLock) error {
+	if err := c.flush(ctx, true, lock); err != nil {
+		return err
+	}
+	// Wait for the changes flushed to be received.
+	if err := c.BlockWait(ctx); err != nil {
+		return err
+	}
+	// Do a final flush
```

Comment:
This catches another bug that the flush generates binlog activities which then causes the compare (`c.canal.SyncedPosition().Compare(c.binlogPosSynced)`) to show changes.

The compare could still show changes from other tables, but it will be less noisy because its guaranteed all changes here are read through.